server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -142,7 +150,17 @@ protected void processInput(DesiredBalanceInput desiredBalanceInput) {
                 );
                 computationsExecuted.inc();
 
-                if (currentDesiredBalance.finishReason() == DesiredBalance.ComputationFinishReason.STOP_EARLY) {
+                final DesiredBalance currentDesiredBalance = currentDesiredBalanceRef.get();
+                if (currentDesiredBalance == DesiredBalance.NOT_MASTER || currentDesiredBalance == DesiredBalance.INITIAL) {
```

Comment:
Sorry with this one, the current desired balance being `INITIAL` couldn't that happen on our first calculation after becoming master?

Reply:
Yeah it can. But whatever comes out from `desiredBalanceComputer.compute(...)` cannot be it. If after `setCurrentDesiredBalance` call, we still get `INITIAL` from `currentDesiredBalanceRef.get()`, the balance must have been changed by a different thread due to becoming master. So we discard the result which should be from an old term.

Reply:
Ah yep, got it. I forgot the `compute` sets the `currentDesiredBalanceRef`